### PR TITLE
Additional pages

### PR DESCRIPTION
**Pull request by GustavoCaso**
@kitop @jgnatch I started with the feature for adding extra pages. #24 

Right now the user can add the extra configuration page to the `initializer`, this will create the extra routes in the `routes.rb` in the `engine`.

Also added an extra generator, to create the `views` and the `controller` so the user can modify the new `actions`, `layout` what ever they want.

I need to add test and polish the code, but for now you can tell me what you think, so we work together on this one.

Thanks.

### CODE REVIEW

**Comment by GustavoCaso** [comment]
@kitop @jgnatch How you doing ?

I was working on the issue, I had some thoughts and I like to know what do you think about.

* Right when the user adds the extra pages configuration, and start the server the `engine` will see them and dynamically will create the routes inside the `engine` routes. I don't know if you like it that way or not.

* Other thing in order for the user to been able to have the views in their application and been able to have what ever content inside them but with the currently working layout of `mountain_view` the new generator  `rails g mountain_view:extra_pages` will have to create the same structure inside the controllers folder of the user with the same content as the one in the gem but with, that way the user can then create the actions and pass what ever the want. So I would like to know if so far is what you guys like for or I should change anything.
Al so the generator will create the views inside the user project following the gem convention `mountain_view > styliguide > extra_page.html.erb`

Thanks for everything.

**Comment by GustavoCaso** [comment]
@kitop @jgnatch any one around for feedback ??

**Comment by kitop** [comment]
Gustavo, sorry for the delay!
I'll try to review it in depth this week!
Thanks for your contribution!!

**Comment by GustavoCaso** [comment]
@kitop.
Thanks is not deep review but hear your opinions on the matter

**Comment by jgnatch** [comment]
Hey @GustavoCaso 
First of all thanks for contributing again and sorry for the delay! 

I just finished checking the code and so far the code looks great.

One thing I'm not completely sure if it makes sense to create a new controller for every static page. It seems like an extra step for the user and I'm not sure if it's worth it.
What do you think of having an extra_pages_controller similar to what's done in this article?
http://blog.teamtreehouse.com/static-pages-ruby-rails#generate

Cheers!

Nacho

cc: @kitop

**Comment by GustavoCaso** [comment]
@jgnatch Ok let me see if I understand what your saying.

Basically the user will add the extra pages in the config block, that will dynamically create inside the `mountain_view` engine with this code https://github.com/GustavoCaso/mountain_view/blob/additional_pages/config/routes.rb#L4

Second the user can use the new rake command to create the views and the controller `rails g mountain_view:extra_pages` this will create the views and the controller if need it.
Just one controller not one for every page(thats the code right now)

We could instead of adding the routes under styleguide resources we could create a new resources call `extra_pages`

```ruby
def extra_routes
  MountainView.configuration.extra_pages.each do |extra_page|
    get "extra_pages/extra_page", as: extra_page
  end
end
```
 That way the controller inside the user application doesn't have to follow our structure of folder, it would only have to use the same layout if they want.

```ruby
class ExtraPagesController < ApplicationController
  layout "mountain_view" # this is optional

  def show
    render template: "pages/#{params[:page]}"
  end
end
```

Is that ok with you guys @jgnatch @kitop ?

**Comment by GustavoCaso** [comment]
@jgnatch @kitop What do you guys think about this one ?

**Comment by GustavoCaso** [comment]
@jgnatch @kitop Anything around here ?

**Comment by kitop** [comment]
@GustavoCaso I like your latest changes!
I think the general idea is good and the code looks good.
Will give it a try in a real application to see how it feels using it. I think we may be able to avoid generating a controller if we already provide the default one. 
Also, if the extra pages controller is going to be namespaced to `MountainView` we can probably remove to option to use a custom name for it, as the chances of clashing are really really low. And that would simplify the code a bit, removing some metaprogramming.

I'll try to give it a test ride this weekend and get back to you.
Thanks for the contribution!! :smile:

**Comment by jgnatch** [comment]
Hey @GustavoCaso really sorry for the delay. I've checked the code and the changes are much better!

As @kitop said I would remove the option for putting a custom name on the controller. As it is namespaced inside MountainView it will not clash with the users' existing controllers and by removing this option the code will be much more simpler in my opinion.

Thanks again!!

**Comment by GustavoCaso** [comment]
Sorry for the long time without response, is been crazy around here, tomorrow I will remove the necessary code to continue with this issue.

Thanks for the patience @jgnatch @kitop

**Comment by jgnatch** [comment]
Hey Gustavo!

Same here!! Lots of stuff lately and now I'm traveling a bit for the
holidays :)

No hurry at all... Thank you, for contributing!!

**Comment by devnacho** [comment]
@GustavoCaso How's it going? Hope you are doing great :)

I would like to merge this PR. Do you want to remove the unnecessary code you mentioned in the last comment or should we merge it and do it ourselves?

Thanks!!!

Nacho

**Comment by GustavoCaso** [comment]
Hey @devnacho long time :smile: 

Sure I will remove it and you can merge it, and we can continue with more issues.

How is everything ?

**Comment by devnacho** [comment]
@GustavoCaso All good here, thanks!!

Trying to be a bit more proactive with mountain view :smile:

**Comment by Frexuz** [comment]
What's the progress here? :D Would really like to see this! <3

**Comment by kitop** [comment]
Hi @Frexuz, thanks for pinging here :)

@GustavoCaso Hi :) Do you want/have time to finish this? If not, I'll can bring the branch locally, change the remaining things and fix the new merge conflicts.

**Comment by GustavoCaso** [comment]
@kitop, @Frexuz thanks for reminding me about it completely forgot about it 😢 

I can clean everything and push update the branch. I think I will have sometime today after work

**Comment by kitop** [comment]
Thank you!! :D

**Comment by GustavoCaso** [comment]
@kitop I updated the code. Woow it was a long time 😆. But it is a really interesting project to work with.

If you like all the changes. I will apply Hound recommendations

**Comment by GustavoCaso** [comment]
@kitop Also the test fail randomly. I do not why. I will look for answer

File: test/generators/extra_pages_generator_test.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+require 'test_helper'
+
+class ExtraPagesGeneratorTest < Rails::Generators::TestCase
+  tests MountainView::Generators::ExtraPagesGenerator
+
+  destination File.expand_path("../../tmp2", __FILE__)
```
Comment (kitop) [inline test/generators/extra_pages_generator_test.rb]: Why do you need to use `tmp2` instead of `tmp` here?

Comment (GustavoCaso) [inline test/generators/extra_pages_generator_test.rb]: Nothing special to have a different place where run the tests. You think it has something . to do with the tests failing?

File: lib/mountain_view/engine.rb [hunk 2]
```diff
@@ -33,6 +33,7 @@ class Engine < ::Rails::Engine
       ActiveSupport.on_load :action_controller do
         helper MountainView::ApplicationHelper
         helper MountainView::ComponentHelper
+        helper MountainView::StyleguideHelper
```
Comment (kitop) [inline lib/mountain_view/engine.rb]: Do we need to include this application-wide? 
It's only used for the styleguide. We can probably put a note in the README that if people want to override the styleguide they can include it themselves.

**Comment by kitop** [comment]
@GustavoCaso thanks for your work!! Just left a couple of comments. The rest looks good to me :)

File: lib/mountain_view/engine.rb (hunk 2 above)
Comment (GustavoCaso) [inline lib/mountain_view/engine.rb]: I just remove it!

File: test/generators/extra_pages_generator_test.rb (hunk 1 above)
Comment (GustavoCaso) [inline test/generators/extra_pages_generator_test.rb]: I will use `tmp` instead

**Comment by GustavoCaso** [comment]
I keep getting some random test failure, but do not know why

File: test/generators/extra_pages_generator_test.rb (hunk 1 above)
Comment (kitop) [inline test/generators/extra_pages_generator_test.rb]: Thanks! Not sure if it has to do with the failing test. Lets see what happens after changing it, and we can investigate if it presists.

**Comment by kitop** [comment]
Mhm, will investigate on the test failure. Thanks!

**Comment by kitop** [comment]
Oh! I think I know why, you may have to set the proper template you're expecting: https://github.com/devnacho/mountain_view/blob/9dcbf5fde4764591af79eef2dab684eab72c7014/test/generators/component_generator_test.rb#L10-L11

If the test runs after the one that changes the template to HAML, it will create a different file (.hmtl.haml instead of .html.erb). Can you try that?

**Comment by kitop** [comment]
Looks great now @GustavoCaso. Just one more thing, can you squash you commits into one or two commits?
Ready to merge after that! Thanks!

**Comment by kitop** [comment]
Released in v0.12.0 🎉

**Pull request merged**